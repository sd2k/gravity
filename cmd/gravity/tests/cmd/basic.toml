bin.name = "gravity"
args = "--world basic --wit-file ../../examples/basic/wit/basic.wit ../../target/wasm32-unknown-unknown/release/example_basic.wasm"
