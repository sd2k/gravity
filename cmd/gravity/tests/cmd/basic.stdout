// Code generated by arcjet-gravity; DO NOT EDIT.

package basic

import "context"
import "errors"
import "github.com/tetratelabs/wazero"
import "github.com/tetratelabs/wazero/api"

import _ "embed"

//go:embed basic.wasm
var wasmFileBasic []byte

type IBasicLogger interface {
	Debug(
		ctx context.Context,
		msg string,
	)
	Info(
		ctx context.Context,
		msg string,
	)
	Warn(
		ctx context.Context,
		msg string,
	)
	Error(
		ctx context.Context,
		msg string,
	)
}

type BasicFactory struct {
	runtime wazero.Runtime
	module wazero.CompiledModule
}

func NewBasicFactory(
	ctx context.Context,
	logger IBasicLogger,
) (*BasicFactory, error) {
	wazeroRuntime := wazero.NewRuntime(ctx)

	_, err0 := wazeroRuntime.NewHostModuleBuilder("arcjet:basic/logger").
	NewFunctionBuilder().
	WithFunc(func(
		ctx context.Context,
		mod api.Module,
		arg0 uint32,
		arg1 uint32,
	) {
		buf0, ok0 := mod.Memory().Read(arg0, arg1)
		if !ok0 {
			panic(errors.New("failed to read bytes from memory"))
		}
		str0 := string(buf0)
		logger.Debug(ctx, str0)
	}).
	Export("debug").
	NewFunctionBuilder().
	WithFunc(func(
		ctx context.Context,
		mod api.Module,
		arg0 uint32,
		arg1 uint32,
	) {
		buf0, ok0 := mod.Memory().Read(arg0, arg1)
		if !ok0 {
			panic(errors.New("failed to read bytes from memory"))
		}
		str0 := string(buf0)
		logger.Info(ctx, str0)
	}).
	Export("info").
	NewFunctionBuilder().
	WithFunc(func(
		ctx context.Context,
		mod api.Module,
		arg0 uint32,
		arg1 uint32,
	) {
		buf0, ok0 := mod.Memory().Read(arg0, arg1)
		if !ok0 {
			panic(errors.New("failed to read bytes from memory"))
		}
		str0 := string(buf0)
		logger.Warn(ctx, str0)
	}).
	Export("warn").
	NewFunctionBuilder().
	WithFunc(func(
		ctx context.Context,
		mod api.Module,
		arg0 uint32,
		arg1 uint32,
	) {
		buf0, ok0 := mod.Memory().Read(arg0, arg1)
		if !ok0 {
			panic(errors.New("failed to read bytes from memory"))
		}
		str0 := string(buf0)
		logger.Error(ctx, str0)
	}).
	Export("error").
	Instantiate(ctx)
	if err0 != nil {
		return nil, err0
	}

	// Compiling the module takes a LONG time, so we want to do it once and hold
	// onto it with the Runtime
	module, err := wazeroRuntime.CompileModule(ctx, wasmFileBasic)
	if err != nil {
		return nil, err
	}
	return &BasicFactory{
		runtime: wazeroRuntime,
		module: module,
	}, nil
}

func (f *BasicFactory) Instantiate(ctx context.Context) (*BasicInstance, error) {
	if module, err := f.runtime.InstantiateModule(ctx, f.module, wazero.NewModuleConfig()); err != nil {
		return nil, err
	} else {
		return &BasicInstance{module}, nil
	}
}

func (f *BasicFactory) Close(ctx context.Context) {
	f.runtime.Close(ctx)
}

type BasicInstance struct {
	module api.Module
}

func (i *BasicInstance) Close(ctx context.Context) error {
	if err := i.module.Close(ctx); err != nil {
		return err
	}

	return nil
}

// writeString will put a Go string into the Wasm memory following the Component
// Model calling convetions, such as allocating memory with the realloc function
func writeString(
	ctx context.Context,
	s string,
	memory api.Memory,
	realloc api.Function,
) (uint64, uint64, error) {
	if len(s) == 0 {
		return 1, 0, nil
	}

	results, err := realloc.Call(ctx, 0, 0, 1, uint64(len(s)))
	if err != nil {
		return 1, 0, err
	}
	ptr := results[0]
	ok := memory.Write(uint32(ptr), []byte(s))
	if !ok {
		return 1, 0, err
	}
	return uint64(ptr), uint64(len(s)), nil
}

func (i *BasicInstance) Hello(
	ctx context.Context,
) (string, error) {
	raw0, err0 := i.module.ExportedFunction("hello").Call(ctx, )
	if err0 != nil {
		var default0 string
		return default0, err0
	}

	// The cleanup via `cabi_post_*` cleans up the memory in the guest. By
	// deferring this, we ensure that no memory is corrupted before the function
	// is done accessing it.
	defer func() {
		if _, err := i.module.ExportedFunction("cabi_post_hello").Call(ctx, raw0...); err != nil {
			// If we get an error during cleanup, something really bad is
			// going on, so we panic. Also, you can't return the error from
			// the `defer`
			panic(errors.New("failed to cleanup"))
		}
	}()

	results0 := raw0[0]
	value1, ok1 := i.module.Memory().ReadByte(uint32(results0 + 0))
	if !ok1 {
		var default1 string
		return default1, errors.New("failed to read byte from memory")
	}
	var value8 string
	var err8 error
	switch value1 {
	case 0:
		ptr2, ok2 := i.module.Memory().ReadUint32Le(uint32(results0 + 4))
		if !ok2 {
			var default2 string
			return default2, errors.New("failed to read pointer from memory")
		}
		len3, ok3 := i.module.Memory().ReadUint32Le(uint32(results0 + 8))
		if !ok3 {
			var default3 string
			return default3, errors.New("failed to read length from memory")
		}
		buf4, ok4 := i.module.Memory().Read(ptr2, len3)
		if !ok4 {
			var default4 string
			return default4, errors.New("failed to read bytes from memory")
		}
		str4 := string(buf4)
		value8 = str4
	case 1:
		ptr5, ok5 := i.module.Memory().ReadUint32Le(uint32(results0 + 4))
		if !ok5 {
			var default5 string
			return default5, errors.New("failed to read pointer from memory")
		}
		len6, ok6 := i.module.Memory().ReadUint32Le(uint32(results0 + 8))
		if !ok6 {
			var default6 string
			return default6, errors.New("failed to read length from memory")
		}
		buf7, ok7 := i.module.Memory().Read(ptr5, len6)
		if !ok7 {
			var default7 string
			return default7, errors.New("failed to read bytes from memory")
		}
		str7 := string(buf7)
		err8 = errors.New(str7)
	default:
		err8 = errors.New("invalid variant discriminant for expected")
	}
	return value8, err8
}

func (i *BasicInstance) Primitive(
	ctx context.Context,
) bool {
	_, err0 := i.module.ExportedFunction("primitive").Call(ctx, )
	// The return type doesn't contain an error so we panic if one is encountered
	if err0 != nil {
		panic(err0)
	}

	results0 := raw0[0]
	value1 := results0 != 0
	return value1
}

func (i *BasicInstance) OptionalPrimitive(
	ctx context.Context,
) (bool, bool) {
	_, err0 := i.module.ExportedFunction("optional-primitive").Call(ctx, )
	// The return type doesn't contain an error so we panic if one is encountered
	if err0 != nil {
		panic(err0)
	}

	results0 := raw0[0]
	value1, ok1 := i.module.Memory().ReadByte(uint32(results0 + 0))
	// The return type doesn't contain an error so we panic if one is encountered
	if !ok1 {
		panic(errors.New("failed to read byte from memory"))
	}
	var result4 bool
	var ok4 bool
	if value1 == 0 {
		ok4 = false
	} else {
		value2, ok2 := i.module.Memory().ReadByte(uint32(results0 + 1))
		// The return type doesn't contain an error so we panic if one is encountered
		if !ok2 {
			panic(errors.New("failed to read byte from memory"))
		}
		value3 := value2 != 0
		ok4 = true
		result4 = value3
	}
	return result4, ok4
}

func (i *BasicInstance) ResultPrimitive(
	ctx context.Context,
) (bool, error) {
	raw0, err0 := i.module.ExportedFunction("result-primitive").Call(ctx, )
	if err0 != nil {
		var default0 bool
		return default0, err0
	}

	// The cleanup via `cabi_post_*` cleans up the memory in the guest. By
	// deferring this, we ensure that no memory is corrupted before the function
	// is done accessing it.
	defer func() {
		if _, err := i.module.ExportedFunction("cabi_post_result-primitive").Call(ctx, raw0...); err != nil {
			// If we get an error during cleanup, something really bad is
			// going on, so we panic. Also, you can't return the error from
			// the `defer`
			panic(errors.New("failed to cleanup"))
		}
	}()

	results0 := raw0[0]
	value1, ok1 := i.module.Memory().ReadByte(uint32(results0 + 0))
	if !ok1 {
		var default1 bool
		return default1, errors.New("failed to read byte from memory")
	}
	var value7 bool
	var err7 error
	switch value1 {
	case 0:
		value2, ok2 := i.module.Memory().ReadByte(uint32(results0 + 4))
		if !ok2 {
			var default2 bool
			return default2, errors.New("failed to read byte from memory")
		}
		value3 := value2 != 0
		value7 = value3
	case 1:
		ptr4, ok4 := i.module.Memory().ReadUint32Le(uint32(results0 + 4))
		if !ok4 {
			var default4 bool
			return default4, errors.New("failed to read pointer from memory")
		}
		len5, ok5 := i.module.Memory().ReadUint32Le(uint32(results0 + 8))
		if !ok5 {
			var default5 bool
			return default5, errors.New("failed to read length from memory")
		}
		buf6, ok6 := i.module.Memory().Read(ptr4, len5)
		if !ok6 {
			var default6 bool
			return default6, errors.New("failed to read bytes from memory")
		}
		str6 := string(buf6)
		err7 = errors.New(str6)
	default:
		err7 = errors.New("invalid variant discriminant for expected")
	}
	return value7, err7
}

